torchrl/data/replay_buffers/samplers.py

```diff
@@ -0,0 +1,179 @@
+            )
+        if beta < 0:
+            raise ValueError(f"beta must be greater or equal to 0, got beta={beta}")
+
+        self._max_capacity = max_capacity
+        self._alpha = alpha
+        self._beta = beta
+        self._eps = eps
+        if dtype in (torch.float, torch.FloatType, torch.float32):
+            self._sum_tree = SumSegmentTreeFp32(self._max_capacity)
+            self._min_tree = MinSegmentTreeFp32(self._max_capacity)
+        elif dtype in (torch.double, torch.DoubleTensor, torch.float64):
+            self._sum_tree = SumSegmentTreeFp64(self._max_capacity)
+            self._min_tree = MinSegmentTreeFp64(self._max_capacity)
+        else:
+            raise NotImplementedError(
+                f"dtype {dtype} not supported by PrioritizedReplayBuffer"
+            )
+        self._max_priority = 1.0
+
+    @property
+    def default_priority(self) -> float:
+        return (self._max_priority + self._eps) ** self._alpha
+
+    def sample(self, storage: Storage, batch_size: int) -> torch.Tensor:
+        p_sum = self._sum_tree.query(0, len(storage))
+        p_min = self._min_tree.query(0, len(storage))
+        if p_sum <= 0:
+            raise RuntimeError("negative p_sum")
+        if p_min <= 0:
+            raise RuntimeError("negative p_min")
+        mass = np.random.uniform(0.0, p_sum, size=batch_size)
+        index = self._sum_tree.scan_lower_bound(mass)
+        if isinstance(index, torch.Tensor):
+            index.clamp_max_(len(storage) - 1)
+        else:
+            index = np.clip(index, None, len(storage) - 1)
+        weight = self._sum_tree[index]
+
+        # Importance sampling weight formula:
+        #   w_i = (p_i / sum(p) * N) ^ (-beta)
+        #   weight_i = w_i / max(w)
+        #   weight_i = (p_i / sum(p) * N) ^ (-beta) /
+        #       ((min(p) / sum(p) * N) ^ (-beta))
+        #   weight_i = ((p_i / sum(p) * N) / (min(p) / sum(p) * N)) ^ (-beta)
+        #   weight_i = (p_i / min(p)) ^ (-beta)
+        # weight = np.power(weight / (p_min + self._eps), -self._beta)
+        weight = np.power(weight / p_min, -self._beta)
+        return index, {"_weight": weight}
+
+    def _add_or_extend(self, index: Union[int, torch.Tensor]) -> None:
+        priority = self.default_priority
+
+        if not (
+            isinstance(priority, float)
+            or len(priority) == 1
+            or len(priority) == len(index)
+        ):
+            raise RuntimeError(
+                "priority should be a scalar or an iterable of the same "
+                "length as index"
+            )
+
+        self._sum_tree[index] = priority
+        self._min_tree[index] = priority
+
+    def add(self, index: int) -> None:
+        super().add(index)
+        self._add_or_extend(index)
+
+    def extend(self, index: torch.Tensor) -> None:
+        super().extend(index)
+        self._add_or_extend(index)
+
+    def update_priority(
```

Comment:
I'd suggest to pay attention to one thing here for update_priority.
For distributed cases, if we do extend and update_priority from different clients, one possible case is the index we want to update is already replaced by a newly appended item, which makes the update incorrect. If the ReplayBuffer is large, the probability can be small so the overall impact should be limited.
In the current version of RLMeta, we added a timestamp to deal with this. https://github.com/facebookresearch/rlmeta/blob/75029c8c1b08f57f19672749efb54f571d9acab8/rlmeta/core/replay_buffer.py#L204
We also have an on-going PR to use global keys instead of buffer indices like reverb in https://github.com/facebookresearch/rlmeta/pull/70. We have more discussion for this design if you like.

Reply:
Thanks for pointing this out.
After I merge the PR I will reference this as an issue. We should probably port TimeStamps to torchrl then